src/network/networkd-dhcp-static-lease.c

```diff
@@ -0,0 +1,236 @@
+        _cleanup_(dhcp_static_lease_free_or_set_invalidp) DHCPStaticLease *current_lease = NULL;
+        union in_addr_union addr;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(userdata);
+
+        network = userdata;
+
+        r = lease_new_static(network, filename, section_line, &current_lease);
+        if (r < 0)
+                return log_oom();
+
+        if (isempty(rvalue)) {
+                current_lease->address.s_addr = 0;
+                TAKE_PTR(current_lease);
+                return 0;
+        }
+
+        r = in_addr_from_string(AF_INET, rvalue, &addr);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse IPv4 address for DHCPv4 static lease, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        r = dhcp_static_lease_set_address(current_lease, &addr.in);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to set IPv4 address for DHCPv4 static lease, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        TAKE_PTR(current_lease);
+
+        return 0;
+}
+
+int config_parse_dhcp_static_lease_hwaddr(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        Network *network;
+        _cleanup_(dhcp_static_lease_free_or_set_invalidp) DHCPStaticLease *current_lease = NULL;
+        struct ether_addr hwaddr;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(userdata);
+
+        network = userdata;
+
+        r = lease_new_static(network, filename, section_line, &current_lease);
+        if (r < 0)
+                return log_oom();
+
+        if (isempty(rvalue)) {
+                current_lease->mac_addr = mfree(current_lease->mac_addr);
+                current_lease->mac_addr_size = 0;
+                return 0;
+        }
+
+        r = ether_addr_from_string(rvalue, &hwaddr);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse MAC address for DHCPv4 static lease, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        r = dhcp_static_lease_set_client_id(current_lease, &hwaddr, ETH_ALEN);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to set MAC address for DHCPv4 static lease, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        current_lease->section->invalid = false;
```

Comment:
Please do not reset the flag here. We should refuse sections which have any invalid settings even if a valid setting overwrites later. This is the current limitation of `section_is_valid()` check, unfortunately.

Reply:
This means if any mac address in a section is invalid, the whole section becomes invalid. Please double check that this is what You inteded with this comment.

Reply:
Right. Unfortunately.